Add `canvas()` function; remove layout/dimension args from `g2()`; add renderer support (#46)
yihui/gglite: examples/canvas.Rmd

---
title: Canvas Options
---

```{r}

> library(gglite)

Reviewer: you failed to notice the caveat when creating this Rmd: since we are using multiple renderers in this doc, we must declare gglite.renderer in the beginning

```suggestion
library(gglite)
# We will use non-default renderers for this document
options(gglite.renderer = 'svg')
```

you should have discovered it through browser testing that the svg plots in this document are blank when rendered in html

```

The `canvas()` function controls chart dimensions, layout spacing, and the

> rendering backend. Pipe it after `g2()` to customize a chart's canvas before

Reviewer: canvas() doesn't need to follow g2() immediately, and i don't want to give users an impression that it does

personally i'd prefer canvas() to appear at the end of the pipe since it should get least attention when we read the code; pls move all canvas() calls to the end in all examples

rendering.

## Dimensions

Use `width` and `height` to set the chart size in pixels. By default (`width =
NULL`), the chart auto-fits to its container width and uses a height of 480 px.

```{r}

> g2(mtcars, hp ~ mpg) |> canvas(width = 500, height = 300)

Reviewer: you failed to follow copilot instructions: create a base plot object, and reuse it throughout the document, instead of repeating it everywhere! 

```suggestion
p = g2(mtcars, hp ~ mpg)
p |> canvas(width = 500, height = 300)
```

then reuse p

i've emphasized this many times: DRY, and don't make readers read irrelevant code; highlight the relevant part

```

## Layout Spacing

The `padding`, `margin`, and `inset` arguments control spacing in pixels. Each
accepts a scalar (all sides) or a length-4 vector `c(top, right, bottom, left)`
where `NA` skips a side.

```{r}
# Add padding around the plot area (useful for parallel coords, etc.)
g2(iris, position = names(iris)[-5], color = ~ Species) |>
  canvas(padding = 30) |>
  coord_parallel() |>

>   legend_color(position = 'bottom')

Reviewer: just reuse p, and show all three arguments, padding/margin/inset

if you can apply different bg colors to each area, that'll be even better, since users will clearly see the padding/margin/inset

```

## Renderer

G2 supports three rendering backends, selected via the `renderer` argument
(case-insensitive):

| Renderer | CDN scripts loaded | Approx. size |
|---|---|---|

> | `"Canvas"` (default) | `g2.min.js` | ~1.1 MB |

Reviewer: ```suggestion
| `"Canvas"` (default and optimized) | `g2.min.js` | ~1.1 MB |
```

then add a row with the render "Canvas" that is not optimized, which uses `@antv/g` + `@antv/g-canvas` + `g2.lite.min.js`; check the total size of them and add to the 3rd column

| `"SVG"` | `@antv/g` + `@antv/g-svg` + `g2.lite.min.js` | ~1.5 MB |
| `"WebGL"` | `@antv/g` + `@antv/g-webgl` + `g2.lite.min.js` | ~1.9 MB |

The default Canvas renderer is the lightest option. SVG output can be inspected
in the browser's DevTools (useful for debugging). WebGL is GPU-accelerated and
suited for very large datasets.

```{r}

> # SVG renderer — standalone preview, no global option needed

Reviewer: move the comment out as a separate paragraph since this is worth emphasizing (no global option needed for standalone preview, i,e. printing a chart in the R console).

g2(mtcars, hp ~ mpg) |> canvas(renderer = 'svg')
```

### Caveat: using SVG or WebGL in a multi-chart document

> `g2.min.js` (Canvas) and `g2.lite.min.js` (SVG/WebGL) **cannot coexist on the

Reviewer: the description is not accurate here: we should explain that g2.min.js is a pre-optimized bundle of g2.lite + g + g-canvas but only supports canvas as the renderer; g2.lite is the core g2 without g or g-canvas and can't be optimized when it's loaded in browser with g or any renderer

so especially "`g2.lite.min.js` (SVG/WebGL)" was not accurate: g2.lite has no svg or webgl renderers, and must be combined with a specific renderer

same HTML page**. In a document that contains multiple charts — such as an R

> Markdown, Quarto, or Jupyter notebook — the CDN scripts are injected once for

Reviewer: also mention Shiny

"injected once" is not accurate for litedown documents, in which the scripts are injected for every plot;

don't mention this technical detail, but just say that setting options(gglite.renderer) explicitly will make sure all plots use g2.lite.min.js, and avoid loading g2.min.js (hence avoiding the collision)

the whole document. If any chart uses SVG or WebGL, all charts must use the
lite bundle.

To switch the entire document to the lite bundle, declare a global renderer
option **at the top of the document** (before any charts are rendered):

```r

> options(gglite.renderer = 'svg')   # or 'webgl'

Reviewer: ```suggestion
options(gglite.renderer = 'svg')   # or 'webgl', 'canvas'
```

```

When this option is set:

> - All charts in the document use `g2.lite.min.js` automatically.

Reviewer: g2.lite plus renderer (g-canvas, g-svg, g-webgl), not only g2.lite

- Individual charts can still override the renderer via `canvas(renderer = ...)`.
  For example, when `options(gglite.renderer = 'svg')` is active, a specific
  chart can use `g2(...) |> canvas(renderer = 'webgl')`.

> For **standalone plots** previewed in the browser (`print()` / RStudio Viewer),

Reviewer: ```suggestion
For **standalone plots** previewed in the browser (printed from the console of an interactive R session),
```

each plot is its own HTML page, so no global option is needed.

> ```{r echo=FALSE}

Reviewer: never do that; just three backticks and r if a code chunk is not to be evaluated

````suggestion
```r
````

# Set global renderer to svg for the rest of this document
options(gglite.renderer = 'svg')
```

```{r}
# With options(gglite.renderer = 'svg') set, all charts use g2.lite
g2(mtcars, hp ~ mpg)
```

```{r}
# Override to WebGL for a specific chart
g2(mtcars, hp ~ mpg) |> canvas(renderer = 'webgl')
```

```{r echo=FALSE}
options(gglite.renderer = NULL)  # restore default
```

## Extra Options

The `...` argument in `canvas()` accepts any additional top-level
`chart.options()` keys (e.g., `clip`, `depth`) that are not covered by the
dedicated parameters.

```{r}

> g2(mtcars, hp ~ mpg) |> canvas(clip = TRUE)

Reviewer: demo the difference between clip and not clip with two plots

```

## WebGL for Large Datasets

The WebGL renderer is GPU-accelerated and handles very large numbers of data
points with less frame-time than Canvas. The example below plots 100,000 random
points to illustrate this:

```{r}
set.seed(42)
n = 100000L

> df = data.frame(x = rnorm(n), y = rnorm(n))

Reviewer: you should try your best to find a online large dataset first; generate data only if your attempts fail, and round data to 2 digits to save space (otherwise the JSON will be huge)
